aten/src/ATen/TensorIterator.cpp

```diff
@@ -799,6 +799,35 @@ void TensorIteratorBase::build_borrowing_binary_float_op(const Tensor& out, cons
         .add_input(b));
 }
 
+void TensorIteratorBase::build_comparison_op(const Tensor& out, const Tensor& a,
+    const Tensor& b) {
+  TensorIteratorConfig config;
+
+  config.set_check_mem_overlap(true);
+  config.add_owned_output(out);
+  config.add_owned_input(a);
+  config.add_owned_input(b);
+  config.allow_cpu_scalars(true);
+  config.promote_inputs_to_common_dtype(true);
+
+  if (!out.defined()) {
```

Comment:
Might be worth a comment here - when the output isn't defined yet, e.g. for functional ops like `a == b`, we want the output to be `bool`. When it is defined, e.g. in `torch.eq(a, b, out=c)`, we don't coerce the output type.